### PR TITLE
[BALDE-27] Event System Second Iteration

### PR DESCRIPTION
**Pull request by Blinkuu**

### CODE REVIEW

File: examples/events_example.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
+//
+// Created by blinku on 02.04.19.
+//
+
+#include <iostream>
+#include "input_manager.h"
+#include "GLFW/glfw3.h"
+#include <mouse_events.h>
```
Comment (BIGbadEL) [inline examples/events_example.cpp]: why <>

File: examples/events_example.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    glfwSetMouseButtonCallback(window, mouse_button_callback);
+    glfwSetCursorPosCallback(window, cursor_position_callback);
+
+    Log::Init();
+
+    EventManager::Subscribe<KeyTypedEvent>(HandleType::SYNC, [](){ std::cout << "KeyTypedEvent!\n"; });
+    while (!glfwWindowShouldClose(window)) {
+
+        InputManager::Update();
+        EventManager::Flush(-1);
```
Comment (BIGbadEL) [inline examples/events_example.cpp]: -1 should be passed by default its weird to make users do something like this

File: src/events/callback/handler.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        if(m_TakenID.empty()) {
+            m_TakenID.push_back(0);
+            m_ID = 0;
+        } else {
+            for(unsigned int i = 1; i < UINT32_MAX; ++i) {
+                if(std::find(m_TakenID.begin(), m_TakenID.end(), i) == m_TakenID.end()) {
+                    m_TakenID.push_back(i);
+                    m_ID = i;
+                    break;
+                }
```
Comment (BIGbadEL) [inline src/events/callback/handler.h:83]: we should handle the situation when "break" didn't happen (when "i" reached UINT32_MAX)

File: src/events/event_manager.h [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                return;
+            }
+        }
+    }
+
+    template<class T, class... Args>
+    void EventManager::Emit(Args&& ... args) {
+        static_assert(std::is_base_of<Event, T>::value, "Event is not the base of T");
+
+        T* event = new T{args...};
```
Comment (BIGbadEL) [inline src/events/event_manager.h:158]: we should stick to one type of initialization.

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

File: examples/events_example.cpp (hunk 1 above)
Comment (Blinkuu) [reply examples/events_example.cpp]: missclick

File: examples/events_example.cpp (hunk 2 above)
Comment (Blinkuu) [reply examples/events_example.cpp]: It is, just made it explicit here. Gonna change it.

File: src/events/event_manager.h (hunk 4 above)
Comment (Blinkuu) [reply src/events/event_manager.h:158]: Then let's proceed with the newest standard

File: src/events/callback/handler.h (hunk 3 above)
Comment (Blinkuu) [reply src/events/callback/handler.h:83]: Done

File: src/events/event_manager.h (hunk 4 above)
Comment (BIGbadEL) [reply src/events/event_manager.h:158]: :(

File: src/events/callback/handler.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    template<class F, class... Args>
+    Handler::Handler(F&& fun, Args&& ... args) {
+        if(m_TakenID.empty()) {
+            m_TakenID.push_back(0);
+            m_ID = 0;
+        } else {
+            bool handled = false;
+            for(unsigned int i = 1; i < UINT32_MAX; ++i) {
+                if(std::find(m_TakenID.begin(), m_TakenID.end(), i) == m_TakenID.end()) {
```
Comment (BIGbadEL) [inline src/events/callback/handler.h:79]: consider if we really need  to search whole vector in every iteration

Comment (Blinkuu) [reply src/events/callback/handler.h:79]: I cannot think of any other solution if we want to optimize IDs usage and we bother about "fragmentation" of IDs.

**Review by BIGbadEL — APPROVED** [review]

**Pull request closed without merging**